### PR TITLE
refactor(snapshots): read file in chunks to calculate checksum

### PR DESCRIPTION
**Pull request by deepthidevaki**
## Description

Reading the whole file to memory to calculate checksum is memory intensive. Instead read the files in chunks and update the checksum on the go.

## Related issues

Relates #6666

## Definition of Done

_Not all items need to be done depending on the issue and the pull request._

Code changes:
* [ ] The changes are backwards compatibility with previous versions
* [ ] If it fixes a bug then PRs are created to [backport](https://github.com/zeebe-io/zeebe/compare/stable/0.24...develop?expand=1&template=backport_template.md&title=[Backport%200.24]) the fix to the last two minor versions. You can trigger a backport by assigning labels (e.g. `backport stable/0.25`) to the PR, in case that fails you need to create backports manually.

Testing:
* [ ] There are unit/integration tests that verify all acceptance criterias of the issue
* [ ] New tests are written to ensure backwards compatibility with further versions
* [ ] The behavior is tested manually
* [ ] The change has been verified by a QA run
* [ ] The impact of the changes is verified by a benchmark 

Documentation: 
* [ ] The documentation is updated (e.g. BPMN reference, configuration, examples, get-started guides, etc.)
* [ ] New content is added to the [release announcement](https://drive.google.com/drive/u/0/folders/1DTIeswnEEq-NggJ25rm2BsDjcCQpDape)

### CODE REVIEW

File: snapshot/src/main/java/io/zeebe/snapshots/impl/SnapshotChecksum.java [hunk 1]
```diff
@@ -72,9 +75,20 @@ private static long createCombinedChecksum(final List<Path> paths) throws IOExce
...
     for (final var path : paths) {
-      if (!path.endsWith(CHECKSUM_FILE_NAME)) {
-        checksum.update(path.getFileName().toString().getBytes(StandardCharsets.UTF_8));
-        checksum.update(Files.readAllBytes(path));
+      final byte[] chunkId = path.getFileName().toString().getBytes(StandardCharsets.UTF_8);
+      checksum.update(chunkId);
+
+      try (final FileChannel channel = FileChannel.open(path, StandardOpenOption.READ)) {
+        final int bufferSize = 1024;
+        final ByteBuffer buff = ByteBuffer.allocate(bufferSize);
```
Comment (npepinpe) [inline snapshot/src/main/java/io/zeebe/snapshots/impl/SnapshotChecksum.java:83]: We don't need to allocate this buffer for every bath - we can simply reuse it. I would also recommend using 4096 as the size, which is the typical page size/block size on most systems. See IoUtil.BLOCK_SIZE for example (or even reuse it :slightly_smiling_face:)

File: snapshot/src/main/java/io/zeebe/snapshots/impl/SnapshotChecksum.java [hunk 2]
```diff
@@ -72,9 +75,20 @@ private static long createCombinedChecksum(final List<Path> paths) throws IOExce
...
+
+        while (channel.position() < channel.size()) {
+          buff.position(0);
+          final int bytesRead = channel.read(buff);
```
Comment (npepinpe) [inline snapshot/src/main/java/io/zeebe/snapshots/impl/SnapshotChecksum.java:87]: AFAIK we can write this more succintly:

```java
buff.clear();
while ((bytesRead = channel.read(buff)) != 0) {
  buff.flip();
  checksum.update(buff);
  buff.clear();
}
```

File: snapshot/src/test/java/io/zeebe/snapshots/impl/SnapshotChecksumTest.java [hunk 3]
```diff
@@ -128,4 +141,24 @@ public void shouldDetectCorruptedSnapshot() throws IOException {
     // then
     assertThat(SnapshotChecksum.verify(corruptedSnapshot)).isFalse();
   }
+
+  @Test
+  public void shouldCalculateSameChecksumOfLargeFile() throws IOException {
+    // given
+    final var largeSnapshot = temporaryFolder.newFolder().toPath();
+    final Path file = largeSnapshot.resolve("file");
+    final String largeData = "a".repeat(1000 * 1000 * 4);
```
Comment (npepinpe) [inline snapshot/src/test/java/io/zeebe/snapshots/impl/SnapshotChecksumTest.java:150]: it's probably fine if it's at least a few times the block size, and it's not a multiple of it, no?

**Review by npepinpe — CHANGES_REQUESTED** [review]
:+1: 

just some minor stuff

File: snapshot/src/main/java/io/zeebe/snapshots/impl/SnapshotChecksum.java (hunk 1 above)
Comment (deepthidevaki) [reply snapshot/src/main/java/io/zeebe/snapshots/impl/SnapshotChecksum.java:83]: > We don't need to allocate this buffer for every bath 

Oops. I don't know why I missed it :see_no_evil: 
Resuing IoUtil.BLOCK_SIZE is a good idea.

File: snapshot/src/main/java/io/zeebe/snapshots/impl/SnapshotChecksum.java [hunk 4]
```diff
@@ -74,20 +75,17 @@ public static boolean verify(final Path snapshotDirectory) throws IOException {
   private static long createCombinedChecksum(final List<Path> paths) throws IOException {
     final Checksum checksum = SnapshotChunkUtil.newChecksum();
 
+    final ByteBuffer buff = ByteBuffer.allocate(IoUtil.BLOCK_SIZE);
```
Comment (npepinpe) [inline snapshot/src/main/java/io/zeebe/snapshots/impl/SnapshotChecksum.java:78]: Should we enforce byte order? Possibly if two nodes have different native byte order, then when comparing checksums they would have different checksums?

Comment (deepthidevaki) [reply snapshot/src/main/java/io/zeebe/snapshots/impl/SnapshotChecksum.java:78]: Does it matter? We are not reading non-byte primitive types.

Comment (npepinpe) [reply snapshot/src/main/java/io/zeebe/snapshots/impl/SnapshotChecksum.java:78]: I don't know how the checksum interprets them - but I guess you're probably right, it most likely doesn't interpret them but just reads byte by byte.

**Review by npepinpe — APPROVED** [review]
:rocket:

**Pull request merged**